### PR TITLE
feat: add bitwise operations to reduceNat? and kernel

### PR DESCRIPTION
**Pull request by joehendrix**
This adds bitwise operations to reduceNat? and the kernel.  It incorporates some basic test cases to validate the correct operations are associated.

### CODE REVIEW

**Comment by joehendrix** [comment]
!bench

**Comment by digama0** [comment]
I want to register my concern with adding more primitives to the kernel while https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.5BRFC.5D.20Checking.20primitive.20types.20in.20the.20kernel/near/408455066 is still unresolved. (Short version: Adding a primitive to the kernel is an assertion that it *definitionally* computes to concrete values on concrete values, which means that for it to be a conservative extension we need the definitions to have enough definitional equalities to justify this. Definitions by well founded recursion are particularly bad here because they do not compute definitionally and I would like to replace them with fuel based definitions or similar.)

**Comment by joehendrix** [comment]
!bench

**Comment by joehendrix** [comment]
I'm going to wait for #3139 to be resolved, and then rebase this on top of that.

I'll extend the test case there so we can validate the bitwise operations in `reduceNat?` let Std revert leanprover/std4#501.

**Comment by semorrison** [comment]
> I want to register my concern with adding more primitives to the kernel while https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.5BRFC.5D.20Checking.20primitive.20types.20in.20the.20kernel/near/408455066 is still unresolved. (Short version: Adding a primitive to the kernel is an assertion that it _definitionally_ computes to concrete values on concrete values, which means that for it to be a conservative extension we need the definitions to have enough definitional equalities to justify this. Definitions by well founded recursion are particularly bad here because they do not compute definitionally and I would like to replace them with fuel based definitions or similar.)
> 
> I am willing to put in the leg work for this but I would like affirmation that this is a desirable direction first.

I'm not the one to decide anything here, but this sounds reasonable and desirable to me! Your short version here was much more helpful to me than the original discussion, in fact. Perhaps converting the short version here to an RFC is the right way to go?

**Comment by joehendrix** [comment]
#3139 was just merged, so I will rebase this onto that, and augment with tests to verify `whnf` and kernel performance improvements.

I think the soundness goals in #188 are good ones, but low priority as it isn't doesn't affect users currently.   These additions don't substantially impact the difficulty of that feature, but the performance of these operations is more likely to affect users.

File: tests/lean/run/lean_nat_bitwise.lean [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+example : 0b1100 ||| 0b1010 = 0b1110 := rfl
+example : let x := 2^20; x ||| x = x := rfl
+
+example : 0      ^^^ 0      = 0 := rfl
+example : 0      ^^^ 0b1111 = 0b1111 := rfl
+example : 0b1111 ^^^ 0      = 0b1111 := rfl
+example : 0b1111 ^^^ 0b1111 = 0      := rfl
+example : 0b1100 ^^^ 0b1010 = 0b0110 := rfl
+example : 0b0110 ^^^ 0b0101 = 0b0011 := rfl
+example : let x := 2^20; x ||| x = x := rfl
```
Comment (semorrison) [inline tests/lean/run/lean_nat_bitwise.lean:27]:
```suggestion
example : let x := 2^20; x ^^^ x = x := rfl
```

File: tests/lean/run/lean_nat_bitwise.lean [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+example : 0      >>> 0      = 0 := rfl
+example : 0b1011 >>> 1      = 0b0101 := rfl
+example : 0b1011 >>> 2      = 0b0010 := rfl
+example : 0      >>> 2^20 = 0 := rfl
```
Comment (semorrison) [inline tests/lean/run/lean_nat_bitwise.lean:32]:
```suggestion
example : 0      >>> 2^20   = 0 := rfl
```

File: tests/lean/run/lean_nat_bitwise.lean [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+example : 0      <<< 0      = 0 := rfl
+example : 0b1011 <<< 1      = 0b010110 := rfl
+example : 0b1011 <<< 2      = 0b101100 := rfl
+example : 0 <<< 2^20 = 0 := rfl
```
Comment (semorrison) [inline tests/lean/run/lean_nat_bitwise.lean:37]:
```suggestion
example : 0      <<< 2^20   = 0 := rfl
```

File: tests/lean/run/lean_nat_bitwise.lean (hunk 1 above)
Comment (digama0) [reply tests/lean/run/lean_nat_bitwise.lean:27]: ???

**Comment by joehendrix** [comment]
!bench

**Pull request merged**